cibuildwheel/__main__.py

```diff
@@ -14,11 +16,15 @@
 )
 from cibuildwheel.util import (
     BuildSelector,
-    Unbuffered
+    Unbuffered,
 )
 
 
-def get_option_from_environment(option_name, platform=None, default=None):
+@overload
+def get_option_from_environment(option_name: str, platform: Optional[str], default: str) -> str: ...  # noqa: E704
+@overload
+def get_option_from_environment(option_name: str, platform: Optional[str] = None, default: None = None) -> Optional[str]: ...  # noqa: E704 E302
```

Comment:
Is this overload redundant? It feels like the one below captures this one

Reply:
The problem then is that `mypy` gets confused that even if you pass a default (and the return value can thus never be `None`), it will complain about assigning an `Optional[str]` to a `str` (e.g, calling `parse_environment(environment_config)` which expects a `str`).

Reply:
yes, I meant that specific line - I'll try it

Reply:
(you probably already know this, but) The way that `@overload` works is really confusing... the signatures marked with @overload are the 'public' signatures of the function, but the last one specifies the types of the arguments inside. So this is equivalent:

```python
@overload
def get_option_from_environment(option_name: str, platform: Optional[str], default: str) -> str: ...  # noqa: E704
@overload
def get_option_from_environment(option_name: str, platform: Optional[str] = None, default: Optional[str] = None) -> Optional[str]: ...  # noqa: E704 E302
def get_option_from_environment(option_name: str, platform: Optional[str] = None, default: Optional[str] = None) -> Optional[str]:  # noqa: E302
    '''
    Returns an option from the environment, optionally scoped by the platform.
```

According to the docs, this is the intended way to use `@overload`:

```python
@overload
def get_option_from_environment(option_name: str, platform: Optional[str], default: str) -> str: ...  # noqa: E704
@overload
def get_option_from_environment(option_name: str, platform: Optional[str] = None, default: Optional[str] = None) -> Optional[str]: ...  # noqa: E704 E302
def get_option_from_environment(option_name, platform=None, default=None):
    '''
    Returns an option from the environment, optionally scoped by the platform.
```

But in this form, `--strict` complains `Function is missing a type annotation`. Without `--strict`, mypy has those params as `Any`, so isn't type-checking the interior. Frustrating.

Reply:
Yes, I tried lots of stuff, when writing that. Also with `Union[]` and generics, but I didn't get anything to work (and pass with `--strict`), so best I could do was what's there.
Very frustrating for such a simple use case, indeed :-(